openapi/spec-generator/specbuilder.go

```diff
@@ -0,0 +1,490 @@
+	// that off for now.
+	b.spec = &Spec{
+		ValidationContext: context.Background(),
+		Model: openapi3.T{
+			OpenAPI: "3.0.3",
+		},
+	}
+
+	if err := b.buildInfo(); err != nil {
+		return nil, err
+	}
+
+	if err := b.buildSecurity(); err != nil {
+		return nil, err
+	}
+
+	if err := b.buildServers(); err != nil {
+		return nil, err
+	}
+
+	if err := b.buildTags(); err != nil {
+		return nil, err
+	}
+
+	if err := b.buildComponentsAndPaths(); err != nil {
+		return nil, err
+	}
+
+	if err := b.spec.Model.Validate(b.spec.ValidationContext); err != nil {
+		return nil, err
+	}
+
+	return b.spec, nil
+}
+
+// buildInfo builds the Info field
+func (b *SpecBuilder) buildInfo() error {
+	if b.spec.Model.Info == nil {
+		b.spec.Model.Info = &openapi3.Info{
+			Version: "1.1.3", // TODO: Schlep this dynamically from VersionInfo
+			Title:   "Nomad",
+			Contact: &openapi3.Contact{
+				Email: "support@hashicorp.com",
+			},
+			License: &openapi3.License{
+				Name: "MPL 2",
+				URL:  "https://github.com/hashicorp/nomad/blob/main/LICENSE",
+			},
+		}
+	}
+	return nil
+}
+
+// buildSecurity builds the Security field
+func (b *SpecBuilder) buildSecurity() error {
+	if b.spec.Model.Security == nil {
+		b.spec.Model.Security = *openapi3.NewSecurityRequirements()
+		b.spec.Model.Security.With(openapi3.SecurityRequirement{
+			"X-Nomad-Token": {},
+		})
+	}
+	return nil
+}
+
+// buildServers builds the Servers field
+func (b *SpecBuilder) buildServers() error {
+	if b.spec.Model.Servers == nil {
+		b.spec.Model.Servers = openapi3.Servers{
+			{
+				Description: "dev-agent",
+				URL:         "{scheme}://{address}:{port}/v1",
+				Variables: map[string]*openapi3.ServerVariable{
+					"address": {Default: "127.0.0.1"},
+					"port":    {Default: "4646"},
+					"scheme":  {Default: "https", Enum: []string{"https", "http"}},
+				},
+			},
+			{
+				Description: "agent",
+				URL:         "{scheme}://{address}:{port}/v1",
+				Variables: map[string]*openapi3.ServerVariable{
+					"address": {Default: "127.0.0.1"},
+					"port":    {Default: "4646"},
+					"scheme":  {Default: "https", Enum: []string{"https", "http"}},
+				},
+			},
+		}
+	}
+	return nil
+}
+
+// buildTags builds the Tags field
+func (b *SpecBuilder) buildTags() error {
+	if b.spec.Model.Tags == nil {
+		b.spec.Model.Tags = openapi3.Tags{}
+	}
+	return nil
+}
+
+// buildComponentsAndPaths builds the Components object graph by parsing the model.
+func (b *SpecBuilder) buildComponentsAndPaths() error {
+	v1API := v1api{}
+	b.spec.Model.Components = openapi3.NewComponents()
+
+	b.spec.Model.Components.RequestBodies = openapi3.RequestBodies{}
+	b.spec.Model.Components.Parameters = openapi3.ParametersMap{}
+	b.spec.Model.Components.Responses = openapi3.NewResponses()
+	b.spec.Model.Components.Schemas = openapi3.Schemas{}
+	b.spec.Model.Components.Headers = openapi3.Headers{}
+	b.spec.Model.Components.SecuritySchemes = openapi3.SecuritySchemes{}
```

Comment:
The repeated Components prefix hinders readability. I'd suggest creating the object first, and then setting, e.g.

```suggestion
	c := openapi3.NewComponents()

	c.RequestBodies = openapi3.RequestBodies{}
	c.Parameters = openapi3.ParametersMap{}
	c.Responses = openapi3.NewResponses()
	c.Schemas = openapi3.Schemas{}
	c.Headers = openapi3.Headers{}
	c.SecuritySchemes = openapi3.SecuritySchemes{}

	b.spec.Model.Components = components
```